### PR TITLE
Add interrupt controller to example se config

### PR DESCRIPTION
**Pull request by takekoputa**
Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

### CODE REVIEW

**Comment by takekoputa** [comment]
I'm not sure what's missing here.
```

```

File: configs/example/se.py [hunk 1]
```diff
@@ -278,5 +278,12 @@ def get_processes(options):
     MemConfig.config_mem(options, system)
     config_filesystem(system, options)
 
+for cpu in system.cpu:
+    cpu.createInterruptController()
+    #cpu.connectAllPorts(system.membus)
+    cpu.interrupts[0].pio = system.membus.master
```
Comment (powerjg) [inline configs/example/se.py:284]: I think you need to do this for each thread. E.g.,

```python
for cpu in system.cpu:
    cpu.createInterruptController()
    for i in range(cpu.numThreads):
        cpu.interrupts[i].pio = system.membus.master
        ....
```

Comment (takekoputa) [reply configs/example/se.py:284]: I still get the same error.

**Comment by powerjg** [comment]
What's the command you're using?

I tried the following and it just worked:

```
build/X86/gem5.opt configs/example/se.py -c 'tests/test-progs/hello/bin/x86/linux/hello;tests/test-progs/hello/bin/x86/linux/hello' --caches --l2cache --l1d_size=32kB --l1i_size=32kB --l2_size=2MB --l1d_assoc=8 --l1i_assoc=8 --l2_assoc=16 --cacheline_size=64 --cpu-type=DerivO3CPU --mem-type=DDR4_2400_8x8 --mem-size=8GB --sys-clock=2.6GHz --cpu-clock=2.6GHz -n 2
```

The following seems to fail:

```
build/X86/gem5.opt configs/example/se.py -c 'tests/test-progs/hello/bin/x86/linux/hello;tests/test-progs/hello/bin/x86/linux/hello' --caches --l2cache --l1d_size=32kB --l1i_size=32kB --l2_size=2MB --l1d_assoc=8 --l1i_assoc=8 --l2_assoc=16 --cacheline_size=64 --cpu-type=DerivO3CPU --mem-type=DDR4_2400_8x8 --mem-size=8GB --sys-clock=2.6GHz --cpu-clock=2.6GHz --smt
```

with the error

```
panic: panic condition !intMasterPort.isConnected() occurred: Int port not connected to anything!
```

**Comment by powerjg** [comment]
This seems to be the problem (from the config.ini):

```
[system.cpu.interrupts0]                                                                                                                                                                                     type=X86LocalApic                                                                                                                                                                                            children=clk_domain                                                                                                                                                                                          clk_domain=system.cpu.interrupts0.clk_domain                                                                                                                                                                 eventq_index=0                                                                                                                                                                                               int_latency=1000                                                                                                                                                                                             pio_latency=100000                                                                                                                                                                                           system=system                                                                                                                                                                                                int_master=system.membus.slave[2]                                                                                                                                                                            int_slave=system.membus.master[1]                                                                                                                                                                            pio=system.membus.master[0]                                                                                                                                                                                                                                                                                                                                                                                               [system.cpu.interrupts0.clk_domain]                                                                                                                                                                          type=DerivedClockDomain                                                                                                                                                                                      clk_divider=16                                                                                                                                                                                               clk_domain=system.cpu_clk_domain                                                                                                                                                                             eventq_index=0                                                                                                                                                                                                                                                                                                                                                                                                            [system.cpu.interrupts1]                                                                                                                                                                                     type=X86LocalApic                                                                                                                                                                                            children=clk_domain                                                                                                                                                                                          clk_domain=system.cpu.interrupts1.clk_domain                                                                                                                                                                 eventq_index=0                                                                                                                                                                                               int_latency=1000                                                                                                                                                                                             pio_latency=100000                                                                                                                                                                                           system=system 
```

**Comment by powerjg** [comment]
Aha! I found the problem. That code isn't used for the classic caches.

You need to update the code in BaseCPU.py: https://github.com/darchr/gem5/blob/master/src/cpu/BaseCPU.py#L205

Have I told you I hate the way this code is structured ;)

**Comment by takekoputa** [comment]
Ahhh....

**Comment by powerjg** [comment]
There are still other problems... I suggest figuring this out in `learning_gem5/part1/simple.py` and then trying to figure out how to apply it to `se.py`

**Comment by self-gautrang** [comment]
Now I got this when I run SMT:
```
command line: build/X86/gem5.opt configs/example/se.py -c 'tests/test-progs/hello/bin/x86/linux/hello;tests/test-progs/hello/bin/x86/
linux/hello' --caches --l2cache --l1d_size=32kB --l1i_size=32kB --l2_size=2MB --l1d_assoc=8 --l1i_assoc=8 --l2_assoc=16 --cacheline_s
ize=64 --cpu-type=DerivO3CPU --mem-type=DDR4_2400_8x8 --mem-size=8GB --sys-clock=2.6GHz --cpu-clock=2.6GHz --smt                     
                           
Global frequency set at 1000000000000 ticks per second
warn: DRAM device capacity (16384 Mbytes) does not match the address range assigned (8192 Mbytes)
0: system.remote_gdb: listening for remote gdb on port 7004
0: system.remote_gdb: listening for remote gdb on port 7005                                                                          
fatal: system.membus has two ports responding within range [0xa000000000000000:0xa000000000001000]:
        system.cpu.interrupts1.int_slave     
        system.cpu.interrupts0.int_slave
```

The CPU config for SMT
```
[system.cpu.interrupts0]
type=X86LocalApic
children=clk_domain
clk_domain=system.cpu.interrupts0.clk_domain
eventq_index=0
int_latency=1000
pio_latency=100000
system=system
int_master=system.membus.slave[2]
int_slave=system.membus.master[1]
pio=system.membus.master[0]

[system.cpu.interrupts1]
type=X86LocalApic
children=clk_domain
clk_domain=system.cpu.interrupts1.clk_domain
eventq_index=0
int_latency=1000
pio_latency=100000
system=system
int_master=system.membus.slave[3]
int_slave=system.membus.master[3]
pio=system.membus.master[2]
```

**Comment by self-gautrang** [comment]
This also breaks `-n 2`:
```
command line: build/X86/gem5.opt configs/example/se.py -c 'tests/test-progs/hello/bin/x86/linux/hello;tests/test-progs/hello/bin/x86/linux/hello' --caches --l2cache --l1d_size=32kB --l1i_size=32kB --l2_size=2MB --l1d_assoc=8 --l1i_assoc=8 --l2_assoc=16 --cacheline_size=64 --cpu-type=DerivO3CPU --mem-type=DDR4_2400_8x8 --mem-size=8GB --sys-clock=2.6GHz --cpu-clock=2.6GHz -n 2

fatal: Port <orphan System>.cpu0.interrupts.pio is already connected to <orphan System>.membus.master[0], cannot connect <orphan System>.membus.master[2]
```

The corresponding config,
```
[system.cpu.interrupts0]
type=X86LocalApic
children=clk_domain
clk_domain=system.cpu.interrupts0.clk_domain
eventq_index=0
int_latency=1000
pio_latency=100000
system=system
int_master=system.membus.slave[2]
int_slave=system.membus.master[1]
pio=system.membus.master[0]


[system.cpu.interrupts1]
type=X86LocalApic
children=clk_domain
clk_domain=system.cpu.interrupts1.clk_domain
eventq_index=0
int_latency=1000
pio_latency=100000
system=system
int_master=system.membus.slave[3]
int_slave=system.membus.master[3]
pio=system.membus.master[2]
```

**Comment by takekoputa** [comment]
Sigh, the GitHub account above is mine. I used the wrong one.

**Pull request closed without merging**